File: signal.dart
Function? _subscriber = null;

class Signal {
  final subscribers = <Function>{};
  dynamic _state;
  Signal(value) {
    _state = value;
  }

  int get value {
    if (_subscriber != null) {
      subscribers.add(_subscriber!);
    }
    return _state;
  }

  set value(var val) {
    _state = val;
    subscribers.forEach((sub) => sub());
  }

  Function effect(Function callback) {
    _subscriber = callback;
    callback();
    _subscriber = null;

    return () => {
      subscribers.remove(callback)
    };
  }
}

void main() {
  final variable = Signal(0);
  
  
  
  final unsubscribe = variable.effect(()=>{
    print('count is ${variable.value}')
  });

  variable.value = 5;
  
  variable.value = 25;
  unsubscribe();
  variable.value = 35;

}

File: signal.php
<?php
$subscriber = null;

class Signal {
    private array $subscribers = [];
    private mixed $state;

    public function __construct($value) {
        $this->state = $value;
    }

    public function __get($name) {
        global $subscriber;
        if ($name === 'value') {
            if ($subscriber !== null) {
                $this->subscribers[] = $subscriber;
            }
            return $this->state;
        }
        return $this->$name;
    }

    public function __set($name, $value) {
        if ($name === 'value') {
            $this->state = $value;
            foreach ($this->subscribers as $sub) {
                $sub();
            }
        }else{
            $this->$name = $value;
        }
    }

    public function effect(callable $callback) :Closure {
        global $subscriber;
        $subscriber = $callback;
        $callback();
        $subscriber = null;

        return function() use ($callback) {
            $key = array_search($callback, $this->subscribers);
            if ($key !== false) {
                unset($this->subscribers[$key]);
            }
        };
    }
}


function signal($value)
{
    return new Signal($value);
}

// Example usage:
$signal = signal(10);

$unsubscribe = $signal->effect(function() use ($signal) {
    echo "Signal value: " . $signal->value . "\n";
});

$signal->value = 20;


$signal->value = 25 ;

$unsubscribe(); // Unsubscribe the callback

$signal->value = 30 ;
